evm/contracts/ZAsset.sol

```diff
@@ -0,0 +1,957 @@
+    }
+
+    function setAttestator(
+        address attestor
+    ) external restricted {
+        ZAssetStorage storage $ = _getZAssetStorage();
+        $.attestor = attestor;
+    }
+
+    function deposit(
+        uint256 amount
+    ) public {
+        ZAssetStorage storage $ = _getZAssetStorage();
+        if ($.underlying == address(0)) revert NotWrapped();
+        IERC20($.underlying).safeTransferFrom(msg.sender, address(this), amount);
+        _mint(msg.sender, amount);
+    }
+
+    function withdraw(
+        uint256 amount
+    ) public {
+        ZAssetStorage storage $ = _getZAssetStorage();
+        if ($.underlying == address(0)) revert NotWrapped();
+        _burn(msg.sender, amount);
+        IERC20($.underlying).transfer(msg.sender, amount);
+    }
+
+    function redeem(
+        uint256[8] calldata proof,
+        uint256[2] calldata commitments,
+        uint256[2] calldata commitmentPok,
+        LightClientSource[] calldata lightClients,
+        uint256 nullifier,
+        uint256 value,
+        address beneficiary,
+        bytes32 attestedMessage,
+        bytes calldata signature
+    ) public {
+        if (lightClients.length > 4) revert TooManyLightClients();
+
+        if (value >= Verifier.R) {
+            revert ValueExceedScalarField();
+        }
+        if (nullifier >= Verifier.R) {
+            revert NullifierExceedScalarField();
+        }
+        if (uint256(attestedMessage) >= Verifier.R) {
+            revert InvalidAttestation();
+        }
+
+        ZAssetStorage storage $ = _getZAssetStorage();
+
+        // Extract withdrawn value before updating
+        uint256 alreadyWithdrawn = $.nullifierBalance[nullifier];
+        if (alreadyWithdrawn + value >= Verifier.R) {
+            revert ValueAfterWithdrawExceedsScalarField();
+        }
+        // Update the withdraw value for the given nullifier
+        $.nullifierBalance[nullifier] += value;
+
+        // Verify attestor knowledge of the deposit/beneficiary link
+        if (
+            !SignatureCheckerLib.isValidSignatureNowCalldata(
+                $.attestor, attestedMessage, signature
+            )
+        ) {
+            revert InvalidAttestationSig();
+        }
+
+        // Extract possible sources
+        LightClientChoice[4] memory lightClientChoices;
+        for (uint256 i = 0; i < lightClients.length; i++) {
+            LightClientSource calldata lightClient = lightClients[i];
+            Counterparty storage counterparty =
+                $.counterparties[lightClient.clientId];
+            if (counterparty.tokenAddressKey == 0) {
+                revert OnlyKnownLightClients();
+            }
+            bytes memory consensusState = IBCStore($.ibcHandler).getClient(
+                lightClient.clientId
+            ).getConsensusState(lightClient.clientId, lightClient.height);
+            (, bytes32 stateRoot,) =
+                abi.decode(consensusState, (uint64, bytes32, bytes32));
```

Comment:
the consensus states are not all necessarily the same shape, we probably want to use the offsets here like we do in the state lens clients (configured per chain)